### ISSUE TITLE
Enhancement: patterndb should allow for subsecond context-timeout

### ISSUE THREAD

**Issue by faxm0dem** [balabit/syslog-ng#369]
We have high traffic for a specific rule which we want to spawn an action on.
However, it is not possible to set a subsecond `context-timeout`, therefore the memory usage of the server grows indefinitely.

Would it be acceptable to change the datatype of this value to `float` or `double`, or is there a workaround you could think of?

**Comment by faxm0dem** [comment]
Thinking this through, maybe it would be a better idea to add a `context-max-size` parameter?

**Comment by bazsi** [comment]
Hmmm, right now the correllation timers are managed with second
resolutions, not trivial to change that.

Max size would be easier, I just dont like the idea from a theoretical
point of view.

How much messages do you see on that rule? How much per context?

**Comment by faxm0dem** [comment]
I worked around this in my production environment by adding a primary key to `context-id`.
And to answer your question: in the original `context-id` I'm seeing 50EPS max per context, and 150EPS on average for the whole rule.

**Comment by bazsi** [comment]
But if you could add something to your context-id, then you were merging
independent contexts, didn't you?

In that case the proper solution was to add that to context-id, and
artificially limiting the amount of messages in a context would be the
workaround. Am I missing something?


-- 
Bazsi

**Comment by faxm0dem** [comment]
Yes I'm splitting contexts. I don't like it but it's the only way not to trigger the oom-killer after a few hours

**Comment by faxm0dem** [comment]
anyway, what do you think would be best here?

**Comment by bazsi** [comment]
Hmm. I was under the impression that adding that value to context-id
actually made sense by separating contexts that were indeed separate
things. If it was a workaround, then I don't really know a good solution,
limiting the amount artificially seems to be a bad solution to me.

Maybe using a non-memory based backend would be useful in such cases, but
that would be _much_ slower.

Anyhow, if the limit would work for you, maybe that's the best way to move
forward, even if my gut feeling is bad about that.

If it's really something that you would like to see, pls change the summary
of this issue.
Thanks.


-- 
Bazsi

**Comment by faxm0dem** [comment]
My gut tells me you will be able to find a solution to my particular problem, which I failed to describe in the first place (sorry about that), so here it is:

I've got a context which is defined by two primary keys. Let's call them `${uid}` and `${path}`. Now, I want to trigger an alert when there are more than `max_msg` messages per `interval` time window.
So what I did was setup `context-id="mycontext-${uid}-${path}"` with `context-timeout="1"` and an action with `condition="$(context-length) > 200`. The problem is, the context timeout is never reached, as there is always at least one message reaching the correlation every second.

**Comment by bazsi** [comment]
In fact, you don't really need the content of messages in the correllation
context, right? It is just a stateful rate limit that works even if the
messages do not follow each other. Am I right to assume that suppress() is
not good enough, as it only works if messages are subsequent to each other?
And, suppress doesn't allow the specification of a new message, it just
spews the same template every time.

We do have a similar use-case in mind already, I just don't know the best
way to move forward (yet).

My rough idea is to extract the stateful part of db-parser(), make its data
structure stand on its own feet. Then syslog-ng could keep message related
state using a common infrastructure. In the db-parser case this state would
be the entire messages, in the case of a stateful rate limit, the state
would just be some counters that allow suppress-like functionality.

**Comment by faxm0dem** [comment]
yes, you are entirely correct.
